### PR TITLE
Split wired-in modules into separate LF packages

### PR DESCRIPTION
**Pull request by cocreature**
This is a first step towards making sure that the package ids for
types defined in daml-prim and daml-stdlib don’t change. This PR
mostly adds all the necessary infrastructure for that and moves
GHC.Types and GHC.Prim to make sure it works.

Until data-dependencies are really solid and we have verified that we
no longer have performance issues with an increasing number of Haskell
packages, we still include the source files in daml-prim and then just
rewrite the references.

We will also need to add tests that these packages really have stable
ids but I’ll leave that for separate PRs since this doesn’t make that
much sense anyway until all of the types have moved to stable packages.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs [hunk 1]
```diff
@@ -177,7 +178,15 @@ getDalfDependencies :: [NormalizedFilePath] -> MaybeT Action (Map.Map UnitId LF.
 getDalfDependencies files = do
     unitIds <- concatMap transitivePkgDeps <$> usesE GetDependencies files
     pkgMap <- useNoFileE GeneratePackageMap
-    pure $ Map.restrictKeys pkgMap (Set.fromList $ map (DefiniteUnitId . DefUnitId) unitIds)
+    let actualDeps = Map.restrictKeys pkgMap (Set.fromList $ map (DefiniteUnitId . DefUnitId) unitIds)
+    -- For now, we unconditionally include all stable packages.
+    -- Given that they are quite small and it is pretty much impossible to not depend on them
+    -- this is fine. We might want to try being more clever here in the future.
+    stablePackages <-
+        fmap (Map.fromList . map (first stableUnitId) . Map.toList) $
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs]: I clearly need more sleep … Thanks! :slightly_smiling_face:

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 2]
```diff
@@ -1369,19 +1377,31 @@ convertModuleName =
...
     unitId <- convertUnitId (envModuleUnitId env) (envPkgMap env) $ GHC.moduleUnitId m
-    pure $ Qualified unitId (convertModuleName $ GHC.moduleName m) x
+    pure $ rewriteQualified env $ Qualified unitId (convertModuleName $ GHC.moduleName m) x
 
 qDA_Types :: Env -> a -> ConvertM (Qualified a)
 qDA_Types env a = do
   pkgRef <- packageNameToPkgRef env "daml-prim"
   pure $ Qualified pkgRef (mkModName ["DA", "Types"]) a
 
+rewriteQualified :: Env -> Qualified a -> Qualified a
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: Good idea, renamed and added a comment.

File: compiler/damlc/stable-packages/src/GenerateStablePackage.hs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  Opts{..} <- execParser (info optParser idm)
+  case optModule of
+    ModuleName ["GHC", "Types"] ->
+      writePackage ghcTypes optOutputPath
+    ModuleName ["GHC", "Prim"] ->
+      writePackage ghcPrim optOutputPath
+    _ -> fail $ "Unknown module: " <> show optModule
+
+lfVersion :: Version
+lfVersion = version1_7
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/stable-packages/src/GenerateStablePackage.hs]: Good point, changed.

File: compiler/damlc/stable-packages/src/GenerateStablePackage.hs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+writePackage :: Module -> FilePath -> IO ()
+writePackage mod path = do
+  let pkg = Package lfVersion (NM.fromList [mod])
+  BS.writeFile path $ encodeArchive pkg
+
+ghcTypes :: Module
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/stable-packages/src/GenerateStablePackage.hs]: oops, forgot to remove it, thanks!

File: compiler/damlc/stable-packages/src/GenerateStablePackage.hs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      , dvalBody = EEnumCon (qual (dataTypeCon dataOrdering)) (mkVariantCon con)
+      }
+    dataProxy = DefDataType
+      { dataLocation= Nothing
+      , dataTypeCon = mkTypeCon ["Proxy"]
+      , dataSerializable = IsSerializable True
+      , dataParams = [(mkTypeVar "a", KStar)]
+      , dataCons = DataRecord []
+      }
+    proxyCtor = DefValue
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/stable-packages/src/GenerateStablePackage.hs]: Looks like we don’t, I had just copied this module but I tried removing them and nothing broke so let’s kill them.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs [hunk 6]
```diff
@@ -145,287 +145,292 @@ packagingTests = testGroup "packaging"
...
-                , bDar
-                ]
-        step "Build migration project"
-        withCurrentDirectory projectUpgrade $
-            callCommandQuiet "daml build --generated-src"
-        assertBool "upgrade-0.0.1.dar was not created" =<< doesFileExist upgradeDar
+    -- Note(MK): The hacks around daml-prim which were already not quite right, e.g.,
+    -- we didn’t include daml-prim from all SDK versions, are broken completely
+    -- now that we split daml-prim into multiple packages. Therefore, we
+    -- disable this for now.
```
[the remark this answers is not in the document]
Comment (cocreature) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs]: I opened https://github.com/digital-asset/daml/issues/3704 and added a warning when you invoke `damlc migrate`.

[the remark this answers is not in the document]
Comment (hurryabit) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs]: The warning is a great idea. 👍

File: compiler/damlc/stable-packages/src/GenerateStablePackage.hs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      , dataCons = DataEnum $ map mkVariantCon cons
+      }
+    valCtor con = DefValue
+      { dvalLocation = Nothing
+      , dvalBinder = (mkVal ("$ctor:" <> con), TCon (qual (dataTypeCon dataOrdering)))
+      , dvalNoPartyLiterals= HasNoPartyLiterals True
+      , dvalIsTest = IsTest False
+      , dvalBody = EEnumCon (qual (dataTypeCon dataOrdering)) (mkVariantCon con)
+      }
+    dataProxy = DefDataType
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/stable-packages/src/GenerateStablePackage.hs]: Very confusing, we didn’t even have a definition for this in `GHC.Types.daml` so it wasn’t actually usable even if you tried. Burned with :fire:

**Pull request merged**